Rover: CONDITION_YAW command to skid-steering Rover

jrcacador opened the pull request:
Related to #29416
Added initial handling logic for the CONDITION_YAW command. An associated autotest was created to validate the new functionality. However, the test currently fails, indicating that further debugging and refinement of the yaw control logic are needed.

peterbarker reviewed (changes requested):
```
AT-0002.7: AP: target not received last 3secs, stopping
```

Will need to cope with that problem in the autotest, you can add a parameter to `wait_heading` to send the attitude request through until the condition is met (or the the call fails....)


The semantics of this in guided mode will need to be determined by @rmackay9 .

*do* we come to zero speed after completing this command or can you compose this with other guided-mode commands that set your vehicle speed?

If we don't compose, *should* we loiter after completing this given we haven't actually given a position command?

Just generally, this is really good work - the structure is correct.

jrcacador commented on Rover/mode.h:
```diff
@@ -574,7 +578,8 @@ class ModeGuided : public Mode
         SteeringAndThrottle,
-        Stop
+        Stop,
+        TurnRateAndHeading
```
Done

jrcacador commented on Rover/mode_guided.cpp:
```diff
@@ -101,6 +101,29 @@ void ModeGuided::update()
+            if (have_attitude_target) {
+                // run steering and throttle controllers
+                calc_steering_to_heading(_desired_yaw_cd, _desired_yaw_rate_cds * 0.01f);
+            } else {
```
Done

jrcacador commented on Rover/mode_guided.cpp:
```diff
@@ -101,6 +101,29 @@ void ModeGuided::update()
+                // run steering and throttle controllers
+                calc_steering_to_heading(_desired_yaw_cd, _desired_yaw_rate_cds * 0.01f);
+            } else {
+                // we have reached the destination so stay here
```
Done

jrcacador commented on Rover/mode_guided.cpp:
```diff
@@ -101,6 +101,29 @@ void ModeGuided::update()
+                    if (!start_loiter()) {
+                        stop_vehicle();
+                    }
+                } else {
```
Done

jrcacador commented:
> Will need to cope with that problem in the autotest, you can add a parameter to wait_heading to send the attitude request through until the condition is met (or the the call fails....)

I don't really understand what I have to do for this, as I was not able to find anything similiar. Am I supposed to create a new parameter for this function? I couldn't really find a way to achieve the intended behaviour with the existing ones.

peterbarker commented:
> > Will need to cope with that problem in the autotest, you can add a parameter to wait_heading to send the attitude request through until the condition is met (or the the call fails....)
> 
> I don't really understand what I have to do for this, as I was not able to find anything similiar. Am I supposed to create a new parameter for this function? I couldn't really find a way to achieve the intended behaviour with the existing ones.

OK, you need to continuously send through or guided thinks you've given up and gone home for the day - that takes 3 seconds.  See `SET_ATTITUDE_TARGET_heading_test_target` - you want to pass in `called_function`

peterbarker commented on Tools/autotest/rover.py:
```diff
@@ -6315,6 +6315,21 @@ def MAV_CMD_DO_CHANGE_SPEED(self):
+        self.wait_ready_to_arm()
+        self.arm_vehicle()
+        self.change_mode('GUIDED')
+        self.run_cmd(
+            mavutil.mavlink.MAV_CMD_CONDITION_YAW,
+            p1=60,  # target angle
+            p2=10,  # degrees/second
+            p3=1,   # -1 is counter-clockwise, 1 clockwise
+            p4=0,   # 1 for relative, 0 for absolute
+        )
+        self.wait_heading(60, called_function=self.SET_ATTITUDE_TARGET_heading_test_target(60, None, 1))
```
```suggestion
        def send_cmd(x, y):
            self.run_cmd(
                mavutil.mavlink.MAV_CMD_CONDITION_YAW,
                p1=60,  # target angle
                p2=10,  # degrees/second
                p3=1,   # -1 is counter-clockwise, 1 clockwise
                p4=0,   # 1 for relative, 0 for absolute
            )
        self.wait_heading(60, called_function=send_cmd)
```

jrcacador replied:
Done

peterbarker commented on Rover/mode_guided.cpp:
```diff
@@ -374,6 +406,31 @@ void ModeGuided::set_desired_turn_rate_and_speed(float turn_rate_cds, float targ
+void ModeGuided::set_desired_turn_rate_and_heading_delta(float yaw_delta_cd, float turn_rate_cds, int8_t direction)
+{
+    /// handle initialisation
+    if (_guided_mode != SubMode::HeadingAndSpeed) {
+        _guided_mode = SubMode::HeadingAndSpeed;
```
Surely this should be using your new SubMode?

peterbarker commented on Rover/mode_guided.cpp:
```diff
@@ -374,6 +406,31 @@ void ModeGuided::set_desired_turn_rate_and_speed(float turn_rate_cds, float targ
+    /// handle initialisation
+    if (_guided_mode != SubMode::HeadingAndSpeed) {
+        _guided_mode = SubMode::HeadingAndSpeed;
+        _desired_yaw_cd = ahrs.yaw_sensor;
```
... this probably shouldn't be setting _desired_yaw_cd here; use a local to hold the variable you will pass through to `set _desired_turn_rate_and_heading`.

... also, not clear this should be within the `else`

jrcacador replied:
Done

jrcacador commented on Rover/mode_guided.cpp:
```diff
@@ -264,6 +294,7 @@ bool ModeGuided::set_desired_speed(float speed)
         return g2.wp_nav.set_speed_max(speed);
     case SubMode::HeadingAndSpeed:
     case SubMode::TurnRateAndSpeed:
+    case SubMode::TurnRateAndHeading:
```
Should I put it next to the ones without speed control for now then? If not, what speed do I set in this mode?

jrcacador replied:
After the changes, the test fails again. But only because I'm not changing the speed. I've tried changing the speed to test the command and it passes the test both for relative and absolute angles. 

That said, all that's left is to know how should this command affect the speed of the vehicle.

jrcacador commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -578,6 +581,29 @@ MAV_RESULT GCS_MAVLINK_Rover::handle_command_nav_set_yaw_speed(const mavlink_com
+            return MAV_RESULT_FAILED;
+        }
+
+        // get final angle, 1 = Relative, 0 = Absolute
+        if (packet.param4 > 0) {
+            // relative angle
+            rover.mode_guided.set_desired_turn_rate_and_heading_delta(packet.param1 * 100.0f, packet.param2 * 100.0f, packet.param3);
+        } else {
+            // absolute angle
+            rover.mode_guided.set_desired_turn_rate_and_heading(packet.param1 * 100.0f, packet.param2 * 100.0f, packet.param3);
+        }
```
I can't do a switch on a float value. So I'm casting it to int to be able to do it...

jrcacador commented on Rover/mode_guided.cpp:
```diff
@@ -168,6 +193,7 @@ float ModeGuided::nav_bearing() const
         return g2.wp_nav.nav_bearing_cd() * 0.01f;
     case SubMode::HeadingAndSpeed:
     case SubMode::TurnRateAndSpeed:
+    case SubMode::TurnRateAndHeading:
```
Done

peterbarker reviewed (changes requested):
Looks good - but please squash all the commits together into a single commit.  Let me know if you're going to have a problem doing that.

jrcacador commented:
Okay, nice I can do that. The support to specify cw and ccw movement isn't implemented yet though. 
I need to do that part and improve the autotest to try all types of parameters for the command.

Should I still add this to this PR and then, when everything's fine, squash everything, or is it better to close this as it is and open a new PR for that part?

IamPete1 commented on Rover/mode_guided.cpp:
```diff
@@ -374,6 +407,32 @@ void ModeGuided::set_desired_turn_rate_and_speed(float turn_rate_cds, float targ
 }
 
+// set desired attitude
+void ModeGuided::set_desired_turn_rate_and_heading(float yaw_angle_cd, float turn_rate_cds, int8_t direction)
```
`direction` is unused.

jrcacador commented on Rover/mode_guided.cpp:
```diff
@@ -374,6 +407,32 @@ void ModeGuided::set_desired_turn_rate_and_speed(float turn_rate_cds, float targ
+        desired_yaw_cd = ahrs.yaw_sensor;
+    }
+
+    set_desired_turn_rate_and_heading(wrap_180_cd(desired_yaw_cd + yaw_delta_cd), turn_rate_cds, direction);
```
I believe that depends on the desired functionality for the command. I agree that basing it on the target heading rather than current sensor heading makes more sense. I just wrote it this way because it was the way implemented in similar functions like `set_desired_heading_delta_and_speed`.

jrcacador commented:
I've implemented support for specifying clockwise or counter-clockwise movement and made a simple test that expects the vehicle to turn to 60º counter-clockwise (performing a -300º rotation)

The test fails and I've tracked with gdb the problem. Further down the intruction flow, in the `AR_AttitudeControl` library, the function `get_turn_rate_from_heading` is called.
This function turns the received angle in radians to an angle in range [-PI, PI] at the beginning, which impossibilitates rotations over 180 degrees, no matter what I do. 

In order to get this to work, I would have to alter this function to instead turn the angle received into an angle in range [-2PI, 2PI]. I've tested this and it worked. However, this modification could break other code that relied on this function. 

So, essentially I'm asking what to do? Do I modify it, nonetheless? Do I create an identical function in the library to use in this scenario only and leave the existing one alone?

IamPete1 commented:
> So, essentially I'm asking what to do? Do I modify it, nonetheless? Do I create an identical function in the library to use in this scenario only and leave the existing one alone?

We have a function to do the conversion for you `wrap_2PI`. It will take the -Pi to Pi and return 0 to 2Pi.

rmackay9 commented on libraries/AP_Math/AP_Math.cpp:
```diff
@@ -257,10 +257,10 @@ ftype wrap_PI(const ftype radian)
 }
 
-ftype wrap_2PI(const ftype radian)
+ftype wrap_2PI(const ftype radian, bool directional)
```
we should probably remove this directional argument because this is a very widely used function

jrcacador replied:
Done

jrcacador commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -578,6 +585,38 @@ MAV_RESULT GCS_MAVLINK_Rover::handle_command_nav_set_yaw_speed(const mavlink_com
+        // param4 : relative offset (1) or absolute angle (0)
+
+        // exit if vehicle is not in Guided mode
+        if (!rover.control_mode->in_guided_mode()) {
```
```
case MAV_CMD_CONDITION_YAW:
    if (rover.g2.motors.have_skid_steering()) {
        return handle_command_condition_yaw(packet);
    }
    // not yet implemented for other vehicle types
    return MAV_RESULT_UNSUPPORTED;
```
Is this check upon receiving the command not enough? Or is it another thing I have to do? From what I understand this function already checks if the vehicle either has skid-steering or is omni.

peterbarker replied:
Move the check into this function - it's where Randy was expecting it :-)

jrcacador replied:
Done

jrcacador commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -578,6 +585,38 @@ MAV_RESULT GCS_MAVLINK_Rover::handle_command_nav_set_yaw_speed(const mavlink_com
+        }
+
+        // Check valid direction parameter
+        int direction = (int) packet.param3;
```
Done

jrcacador commented:
@rmackay9 Friendly reminder of this PR :)

jrcacador commented:
@rmackay9 Is there anything else to change?

peterbarker commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -500,6 +500,13 @@ MAV_RESULT GCS_MAVLINK_Rover::handle_command_int_packet(const mavlink_command_in
     switch (packet.command) {
 
+    case MAV_CMD_CONDITION_YAW:
+        if (rover.g2.motors.have_skid_steering()) {
```
We probably should have a "can_pivot_turn()" here, but this is fine.  I don't think omnis should be returning "true" for have_skid_steering as there are many ways to do omnis that aren't skid steering

Not your problem to fix :-)

peterbarker commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -578,6 +585,38 @@ MAV_RESULT GCS_MAVLINK_Rover::handle_command_nav_set_yaw_speed(const mavlink_com
+
+        // Check valid direction parameter
+        int16_t direction = static_cast<int16_t>(packet.param3);
+        if (direction != 0 && direction != 1 && direction != -1) {
```
```suggestion
        if (abs(direction) > 1) {  // only 0, -1 and 1 are valid direction values
```

Very much optional change.

jrcacador replied:
Way more elegant :) . Done

peterbarker commented on Rover/mode_guided.cpp:
```diff
@@ -374,6 +407,38 @@ void ModeGuided::set_desired_turn_rate_and_speed(float turn_rate_cds, float targ
+}
+
+// set desired attitude
+void ModeGuided::set_desired_turn_rate_and_heading_delta(float yaw_delta_cd, float turn_rate_cds, int16_t direction)
```
Why `int16_t`?

jrcacador replied:
I was using `int` previously but @rmackay9 said in a conversation above that `int` is not normally used and suggested the change to `int16_t`.

> we don't use int normally. perhaps int16_h

peterbarker commented on Rover/mode_guided.cpp:
```diff
@@ -374,6 +407,38 @@ void ModeGuided::set_desired_turn_rate_and_speed(float turn_rate_cds, float targ
+{
+    /// handle initialisation
+    float desired_yaw_cd = _desired_yaw_cd;
+    if (_guided_mode != SubMode::TurnRateAndHeading) {
+        _guided_mode = SubMode::TurnRateAndHeading;
+    }
```
```suggestion
    _guided_mode = SubMode::TurnRateAndHeading;
```

jrcacador replied:
Done

peterbarker commented on Tools/autotest/rover.py:
```diff
@@ -6292,6 +6292,33 @@ def MAV_CMD_DO_CHANGE_SPEED(self):
+                    p1=heading,     # target angle
+                    p2=10,          # degrees/second
+                    p3=direction,   # -1 is counter-clockwise, 1 clockwise
+                    p4=0,           # 1 for relative, 0 for absolute
```
Could we get some tests for relative too, please?

peterbarker commented on Rover/mode_guided.cpp:
```diff
@@ -101,6 +101,30 @@ void ModeGuided::update()
+        {
+            // stop vehicle if target not updated within 3 seconds
+            if (have_attitude_target && (millis() - _des_att_time_ms) > 3000) {
+                gcs().send_text(MAV_SEVERITY_WARNING, "target not received last 3secs, stopping");
```
```suggestion
                GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "target not received last 3secs, stopping");
```

peterbarker commented on Tools/autotest/rover.py:
```diff
@@ -6292,6 +6292,33 @@ def MAV_CMD_DO_CHANGE_SPEED(self):
+            """check if rotating in correct direction"""
+            self.wait_heading(intermediate_heading-0.5, intermediate_heading+0.5, called_function=send_cmd)
+            """target angle"""
+            self.wait_heading(heading-0.5, heading+0.5, called_function=send_cmd)
```
Add a line here to make sure that we end up with a roughly zero rate at roughly our target heading...

jrcacador replied:
Done

peterbarker commented on libraries/AP_Math/AP_Math.cpp:
```diff
@@ -266,6 +266,11 @@ ftype wrap_2PI(const ftype radian)
     return res;
 }
 
+ftype wrap_neg2PI_2PI(const ftype radian)
```
We like to have tests for all of our functions in AP_Math - it's the gtest stuff in libraries/AP_Math/tests and isn't hard to work with

This implementation should also go in the header to save some bytes

peterbarker commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -578,6 +581,43 @@ MAV_RESULT GCS_MAVLINK_Rover::handle_command_nav_set_yaw_speed(const mavlink_com
+    }
+
+    // Check valid direction parameter
+    int16_t direction = static_cast<int16_t>(packet.param3);
```
```suggestion
    const int16_t direction = static_cast<int16_t>(packet.param3);
```

... 'though not sure why you went with int16_t

peterbarker reviewed (approved):
LGTM

jrcacador committed:
AP_Math: created new wrap_neg2PI_2PI function
Created a new function that wraps an angle in radians to -2PI..2PI
Created a gtest in AP_Math/tests that verifies the correct behavior of the function.

Co-authored-by: Tiago Santos <tiago.castro.santos@tecnico.ulisboa.pt>

jrcacador committed:
AR_AttitudeControl: created new functions
Added new direction version of get_steering_out_heading and get_turn_rate_from_heading.
The new functions allow bigger rotations up to 360º in either direction,
while the already existing ones only allow up to 180º in either direction

Co-authored-by: Tiago Santos <tiago.castro.santos@tecnico.ulisboa.pt>

jrcacador committed:
Rover: Implemented CONDITION_YAW for skid-steering Rover
Added support for handling MAV_CMD_CONDITION_YAW in Rover (limited to skid-steering or omni vehicles).
Added new function calc_directional_steering_to_heading to allow bigger rotations up to 360º in either direction.
Created two tests (absolute and relative angles) with multiple calls of the command and different sets of
parameters that verify the correct behavior of the command.

Co-authored-by: Tiago Santos <tiago.castro.santos@tecnico.ulisboa.pt>

peterbarker reviewed (changes requested):
I've just realised that the command you're using here isn't actually the right command!

See Copter's "within_distance" stuff to see what these are supposed to be for.

This is supposed to delay the state machine until you reach a yaw - not actually command the vehicle to achieve that yaw.  Copter also implements this command in the wrong way.

This should be done via `set_attitude_target`

rmackay9 commented:
Hi @peterbarker, yes, indeed.  In Copter's defense though, there was no mission command that could command the yaw so we used this command to give us that ability.  Actually I think there still is no mission command to move just the yaw.
